### ISSUE TITLE
#10 [Model] : create ML models for recommendation using NLP

### ISSUE THREAD

**Issue by prathikshetty2002** [tcet-opensource/resume-screener#28]
- [x] created ML model in models/ for skill_recommendation
- [x] created ML model in models/ for course_recommendation

**Comment by prathikshetty2002** [comment]
@yashgosa do this modelling part

**Comment by yashgosa** [comment]
@prathikshetty2002 I have cloned the repo and created a issue specific branch and pulled the code from staging branch.
Is there any Thakur resume related data available if not how should I proceed forward ?

On doing some research I found `Doc2Vec` model and `PyPDF2` to be quite useful but still I will need data to work on

Also found these three datasets useful for future use:
1. [shreya resume dataset](https://www.kaggle.com/datasets/shreyasdhurde/shreya-resume-datasetshreya-resume-dataset)
2. [resume dataset by palaksood97](https://www.kaggle.com/datasets/palaksood97/resume-dataset)
3. [resume dataset by aishikai](https://www.kaggle.com/datasets/aishikai/resume-dataset?select=pdf)

**Comment by yashgosa** [comment]
@Hrishikesh332 

> @prathikshetty2002 I have cloned the repo and created a issue specific branch and pulled the code from staging branch. Is there any Thakur resume related data available if not how should I proceed forward ?
> 
> On doing some research I found `Doc2Vec` model and `PyPDF2` to be quite useful but still I will need data to work on
> 
> Also found these three datasets useful for future use:
> 
> 1. [shreya resume dataset](https://www.kaggle.com/datasets/shreyasdhurde/shreya-resume-datasetshreya-resume-dataset)
> 2. [resume dataset by palaksood97](https://www.kaggle.com/datasets/palaksood97/resume-dataset)
> 3. [resume dataset by aishikai](https://www.kaggle.com/datasets/aishikai/resume-dataset?select=pdf)

@Hrishikesh332 ??

**Comment by prathikshetty2002** [comment]
> 

nice work we can use these resumes for general format currently we are focusing on thakur resume format. i have asked for the data however we will be doing the keyword based recommendation so I suggest create/find a dataset of resume keywords for skills and then perform content based recommendation - eg: if a person is from web domain recommends web dev skills

**Comment by yashgosa** [comment]
- [ ] Data Collection: Collect resumes from different sources in a structured format, such as CSV or JSON from kaggle. [ *add the links to the dataset ( make sure the data is tech related )* ]

- [ ] Preprocessing: Clean and preprocess the resumes to remove any irrelevant information such as contact details, address, etc. Use NLP techniques such as tokenization, stop-word removal, stemming, and lemmatization to convert the text data into a format that can be used for training.

- [ ] Feature Extraction: Use feature extraction techniques such as Bag-of-Words, TF-IDF, and Word Embeddings to convert the preprocessed text data into numerical vectors.

- [ ] Training: Train the ML model using different algorithms such as Random Forest, SVM, or Neural Networks. Label the resumes with the relevant skills associated with the job position.

- [ ] Evaluation: Evaluate the model's performance using different metrics such as accuracy, precision, recall, and F1-score.

- [ ] Recommendation : Recommend the courses and skills from the pdf provided by @prathikshetty2002 

*in future*

- [ ] Deployment: Deploy the trained model as an API that can take in a resume as input and return the recommended skills based on the job position.

- [ ] Optimization: Continuously optimize the model by collecting feedback from users and retraining the model based on the feedback.

- [ ] Integration: Integrate the model with existing HR systems such as Applicant Tracking Systems (ATS) to provide real-time skill recommendations to recruiters and hiring managers.

**Comment by prathikshetty2002** [comment]
good work @yashgosa Phase-1 is about recommending the skills and courses based on user resume

**Comment by yashgosa** [comment]
@prathikshetty2002 @Hrishikesh332 

I have completed both of the models, but am facing issue in saving its output in JSON format. Currently I am saving it like:

    with open("output.json", "w") as f:
        json.dump(data, f, indent=4)

but its overwriting its previous version due to which I cant save `recommended_courses` and `recommended_skills ` in same JSON file.

Also @prathikshetty2002 please share the resource for converting PDF files in JSON files. Currently I am taking inputs from self-built JSON file

**Comment by prathikshetty2002** [comment]
@yashgosa inside models prepare a folder for model_data and inside that add all json files u have used as it gets messy otherwise

**Comment by yashgosa** [comment]
@prathikshetty2002 @Hrishikesh332 please review

**Comment by prathikshetty2002** [comment]
> @prathikshetty2002 @Hrishikesh332 please review
> 

Could u attach the output screenshots here in the issue so we can see them.

**Comment by yashgosa** [comment]
![image](https://user-images.githubusercontent.com/113308290/232671320-c4b2b25c-925c-4f11-8bd1-c89f140d907e.png)
![image](https://user-images.githubusercontent.com/113308290/232671473-c5f29ac2-7761-4088-b093-6ae0322ac5ac.png)
![image](https://user-images.githubusercontent.com/113308290/232671539-683e026d-744d-4e05-a4dd-6a277b6fa814.png)

**Comment by yashgosa** [comment]
When the skill or course is not there in the DB I am adding "not available in database" also we need to find skill related datasets if not present I think so we can create a ggl form of and ask them to enter their skills and add description of it.

**Comment by yashgosa** [comment]
@prathikshetty2002 @Hrishikesh332

**Comment by prathikshetty2002** [comment]
> @prathikshetty2002 @Hrishikesh332 

Can you show the output it is giving on my resume json data?

**Comment by yashgosa** [comment]
can you send me your resume also I am struggling with conversion of pdf to json so, I will have to manually do it...

**Comment by yashgosa** [comment]
@prathikshetty2002

**Comment by yashgosa** [comment]
{
    "similar_courses": {
        "machine learning": [
            "introduction to graduate algorithms",
            "intro to machine learning with pytorch",
            "intro to machine learning with tensorflow",
            "intro to html and css"
        ],
        "machine": [
            "data engineering with aws",
            "oracle cloud infrastructure foundations",
            "oracle cloud infrastructure architect professional",
            "oracle cloud infrastructure architect associate",
            "oracle autonomous database administration"
        ],
        "learning": [
            "data engineering with aws",
            "time series forecasting",
            "segmentation and clustering",
            "product design",
            "classification models"
        ],
        "nosql basic": [
            "data engineering with aws",
            "getting started with google workspace",
            "data analysis and visualization",
            "spark",
            "sql for data analysis"
        ],
        "nosql": [
            "design of computer programs",
            "mobile design and usability for android",
            "google maps apis",
            "how to make an ios app",
            "firebase analytics: android"
        ],
        "basic": [
            "design of computer programs",
            "mobile design and usability for android",
            "google maps apis",
            "how to make an ios app",
            "firebase analytics: android"
        ],
        "python development": [
            "data engineering with aws",
            "product design",
            "classification models",
            "problem solving with advanced analytics",
            "creating an analytical dataset"
        ],
        "python": [
            "data engineering with aws",
            "oracle autonomous database administration",
            "managing remote teams with upwork",
            "get your startup started",
            "how to build a startup"
        ],
        "development": [
            "data engineering with aws",
            "getting started with google workspace",
            "data analysis and visualization",
            "spark",
            "sql for data analysis"
        ],
        "embedded systems": [
            "full stack javascript developer",
            "introduction to tensorflow lite",
            "big data analytics in healthcare",
            "flying car and autonomous flight engineer"
        ],
        "embedded": [
            "data engineering with aws",
            "time series forecasting",
            "segmentation and clustering",
            "product design",
            "classification models"
        ],
        "systems": [
            "data engineering with aws",
            "oracle cloud infrastructure foundations",
            "oracle cloud infrastructure architect professional",
            "oracle cloud infrastructure architect associate",
            "oracle autonomous database administration"
        ],
        "microsoft azure fundamentals": [
            "data engineering with aws",
            "introduction to digital project management",
            "data analysis and visualization",
            "spark",
            "intro to inferential statistics"
        ],
        "microsoft": [
            "data engineering with aws",
            "app marketing",
            "time series forecasting",
            "segmentation and clustering",
            "product design"
        ],
        "azure": [
            "design of computer programs",
            "mobile design and usability for android",
            "google maps apis",
            "how to make an ios app",
            "firebase analytics: android"
        ],
        "fundamentals": [
            "data engineering with aws",
            "firebase analytics: android",
            "network security",
            "intro to backend",
            "Algorithms, Part I - Princeton"
        ]
    }
}



sample output for course recc

**Comment by yashgosa** [comment]
@prathikshetty2002 @Hrishikesh332 I wanted to modify pyresparser  so that it could extract courses also, but how can I do it? I clone it make the required changes but then what. If I have to use pyresparser I will have to pip install it and then import it. But how will I do it if the updated version is in my local machine? Hope so you are able to understand my issue.

**Comment by prathikshetty2002** [comment]
done NLP modelling part
